facilitator/src/transport/gcs.rs

```diff
@@ -0,0 +1,601 @@
+use crate::{
+    config::GCSPath,
+    transport::{Transport, TransportWriter},
+    Error,
+};
+use anyhow::{anyhow, Context, Result};
+use chrono::{prelude::Utc, DateTime, Duration};
+use serde::Deserialize;
+use std::{
+    io::{Cursor, Read, Write},
+    mem,
+};
+
+const STORAGE_API_BASE_URL: &str = "https://storage.googleapis.com";
+const DEFAULT_OAUTH_TOKEN_URL: &str =
+    "http://metadata.google.internal:80/computeMetadata/v1/instance/service-accounts/default/token";
+
+// A wrapper around an Oauth token and its expiration date.
+#[derive(Debug)]
+struct OauthToken {
+    token: String,
+    expiration: DateTime<Utc>,
+}
+
+impl OauthToken {
+    // Returns true if the token is not yet expired.
+    fn expired(&self) -> bool {
+        Utc::now() < self.expiration
+    }
+}
+
+// Represents the response from a GET request to the GKE metadata service's
+// service account token endpoint. Structure is derived from empirical
+// observation of the JSON scraped from inside a GKE job.
+#[derive(Debug, Deserialize, PartialEq)]
+struct MetadataServiceTokenResponse {
+    access_token: String,
+    expires_in: i64,
+    token_type: String,
+}
+
+// Represents the response from a POST request to the GCP IAM service's
+// generateAccessToken endpoint.
+// https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/generateAccessToken
+#[derive(Debug, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+struct GenerateAccessTokenResponse {
+    access_token: String,
+    expire_time: DateTime<Utc>,
+}
+
+// OauthTokenProvider manages a default service account Oauth token (i.e. the
```

Comment:
This should document the allowable values for these Options. Can they all be None? Are there any relationships such that when one of these Options is Some we expect another of them to be Some?